pubsub/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -119,9 +134,16 @@ def ack(self, items):
             if time_to_ack is not None:
                 self._manager.ack_histogram.add(time_to_ack)
 
-        ack_ids = [item.ack_id for item in items]
-        request = types.StreamingPullRequest(ack_ids=ack_ids)
-        self._manager.send(request)
+        # We must potentially split the request into multiple smaller requests
+        # to avoid the server-side max request size limit.
+        ack_ids = (item.ack_id for item in items)
+        total_chunks = int(math.ceil(len(items) / _ACK_IDS_BATCH_SIZE))
+
+        for _ in range(total_chunks):
+            request = types.StreamingPullRequest(
+                ack_ids=itertools.islice(ack_ids, _ACK_IDS_BATCH_SIZE)
```

Comment:
Wont this send only the first _ACK_IDS_BATCH_SIZE ack-ids |total-chunks|-times? It wont actually result in the sending of all of the ack ids, only the first _ACK_IDS_BATCH_SIZE.

Reply:
Since `ack_ids` is a generator, iterating over a slice when building a `StreamingPullRequest` instance consumes the first `_ACK_IDS_BATCH_SIZE` items. Then in the following iteration the _next_ `_ACK_IDS_BATCH_SIZE` items are consumed.

```py
>>> from itertools import islice
>>> BATCH_SIZE = 3
>>> gen = (x for x in range(8))
>>> list(islice(gen, BATCH_SIZE))
[0, 1, 2]
>>> list(islice(gen, BATCH_SIZE))
[3, 4, 5]
>>> list(islice(gen, BATCH_SIZE))
[6, 7]
>>> list(islice(gen, BATCH_SIZE))
[]
```
Iterating over a generator has the benefit of automatically advancing the generator, sparing us the need to keep track of the slice boundaries manually.

Reply:
Oh, geez. I see. Thanks!